TcUnit-Runner/VisualStudioInstance.cs

```diff
@@ -107,29 +107,56 @@ public void Close()
             loaded = false;
         }
 
-        private void LoadDevelopmentToolsEnvironment(string visualStudioVersion)
+        private void LoadDevelopmentToolsEnvironment(string visualStudioVersion, bool pinnedTCVersion)
         {
             /* Make sure the DTE loads with the same version of Visual Studio as the
              * TwinCAT project was created in
              */
             
             // Load the DTE
             string VisualStudioProgId = VisualStudioDteAvailable(visualStudioVersion);
-
+            bool versionAvailable = false;
             dte.UserControl = false; // have devenv.exe automatically close when launched using automation
             dte.SuppressUI = true;
             // Make sure all types of errors in the error list are collected
             dte.ToolWindows.ErrorList.ShowErrors = true;
             dte.ToolWindows.ErrorList.ShowMessages = true;
             dte.ToolWindows.ErrorList.ShowWarnings = true;
+            // First set the SilentMode and then try to open the Remote Manager
+            var tcAutomationSettings = dte.GetObject("TcAutomationSettings");
+            tcAutomationSettings.SilentMode = false; // Only available from TC3.1.4020.0 and above
 
             // Load the correct version of TwinCAT using the remote manager in the automation interface
             log.Info("Using the TwinCAT remote manager to load TwinCAT version '" + this.tcVersion + "'...");
             ITcRemoteManager remoteManager = dte.GetObject("TcRemoteManager");
-            remoteManager.Version = this.tcVersion;
+            
+            //Check if the Version is Pinned, if the Version is Pinned check if the Pinned Version is available
+            if (pinnedTCVersion)
+            {
+                log.Info("Project has a pinned Version : " + this.tcVersion);
+                foreach (var possibleVersion in remoteManager.Versions)
+                {
+                   if (possibleVersion == this.tcVersion)
+                    {
+                      versionAvailable = true;
+                      break;
+                    }
+                }
+                if (versionAvailable)
+                {
+                    log.Info("The pinned TC Version is available ");
+                } else
+                {
+                    log.Error("The pinned TC Version is not available ");
+                    throw new Exception("The pinned TC Version is not available");
+                }
+            }
 
-            var tcAutomationSettings = dte.GetObject("TcAutomationSettings");
-            tcAutomationSettings.SilentMode = true; // Only available from TC3.1.4020.0 and above
+            if (!pinnedTCVersion || versionAvailable)
```

Comment:
I'm not sure I follow this. What happens if we don't pin, and load a project in where the version of TwinCAT is not available?

Reply:
I think the name "versionAvailable" is not meaningful here.
"TcVersionIsPinnedAndVersionAvailable"

These things happen.
The first thing to check is whether the TwinCat Project is pinned.
If it is pinned it will be checked if the version is available.

Now comes 
`if (!pinnedTCVersion || versionAvailable)
            {
                remoteManager.Version = this.tcVersion;
                log.Info("Load TwinCAT version: "+ remoteManager.Version);
            }`
If the project is not pinned or it is pinned and the pinned version is available

It starts the RemoteManger as before
Hint: (If the version is not available and the project is not pinned, the Remote Manager will start in the latest version)

It is only aborted if the project is pinned and the version is not available.

Reply:
Makes total sense. Thanks for clarification.